client-spark/shuffle-manager-3/src/main/scala/org/apache/spark/sql/execution/columnar/RssCompressionSchemes.scala

```diff
@@ -236,3 +239,189 @@ case object RssPassThrough extends RssCompressionScheme {
+    case RSS_INT | RSS_LONG | RSS_STRING => true
+    case _ => false
+  }
+
+  class RssEncoder[T <: AtomicType](columnType: NativeRssColumnType[T]) extends Encoder[T] {
+    // Size of the input, uncompressed, in bytes. Note that we only count until the dictionary
+    // overflows.
+    private var _uncompressedSize = 0
+
+    // If the number of distinct elements is too large, we discard the use of dictionary encoding
+    // and set the overflow flag to true.
+    var overflow = false
+
+    // Total number of elements.
+    private var count = 0
+
+    def cleanBatch: Unit = {
+      count = 0
+      _uncompressedSize = 0
+    }
+
+    // The reverse mapping of _dictionary, i.e. mapping encoded integer to the value itself.
+    private var values = new mutable.ArrayBuffer[T#InternalType](1024)
+
+    // The dictionary that maps a value to the encoded short integer.
+    private val dictionary = new java.util.HashMap[Any, Short](1024)
+
+    // Size of the serialized dictionary in bytes. Initialized to 4 since we need at least an `Int`
+    // to store dictionary element count.
+    private var dictionarySize = 4
+
+    override def gatherCompressibilityStats(row: InternalRow, ordinal: Int): Unit = {
+      if (!overflow) {
+        val value = columnType.getField(row, ordinal)
+        val actualSize = columnType.actualSize(row, ordinal)
+        count += 1
+        _uncompressedSize += actualSize
+        if (!dictionary.containsKey(value)) {
+          if (dictionary.size < MAX_DICT_SIZE) {
+            val clone = columnType.clone(value)
+            values += clone
+            dictionarySize += actualSize
+            dictionary.put(clone, dictionary.size.toShort)
+          } else {
+            overflow = true
+            values.clear()
+            dictionary.clear()
+          }
+        }
+      }
+    }
+
+    override def compress(from: ByteBuffer, to: ByteBuffer): ByteBuffer = {
+      to.putInt(RssDictionaryEncoding.typeId)
+        .putInt(dictionary.size)
+
+      var i = 0
+      while (i < values.length) {
+        columnType.append(values(i), to)
+        i += 1
+      }
+
+      while (from.hasRemaining) {
+        to.putShort(dictionary.get(columnType.extract(from)))
+      }
+
+      to.rewind()
+      to
+    }
+
+    override def uncompressedSize: Int = _uncompressedSize
+
+    override def compressedSize: Int = if (overflow) Int.MaxValue else dictionarySize + count * 2
+  }
+
+  class RssDecoder[T <: AtomicType](buffer: ByteBuffer, columnType: NativeRssColumnType[T])
+    extends Decoder[T] {
+    val elementNum = ByteBufferHelper.getInt(buffer)
+    private val dictionary: Array[Any] = new Array[Any](elementNum)
+    private var intDictionary: Array[Int] = null
+    private var longDictionary: Array[Long] = null
+    private var stringDictionary: Array[String] = null
+
+    columnType.dataType match {
+      case _: IntegerType =>
+        intDictionary = new Array[Int](elementNum)
```

Comment:
Codes in the three branches are nearly identical, I think we'd better extract common logic to be more concise.